src/libecalc/core/models/compressor/train/variable_speed_compressor_train_common_shaft_multiple_streams_and_pressures.py

```diff
@@ -405,6 +407,70 @@ def get_max_standard_rate_per_stream(
+        self,
+        suction_pressure: float,
+        target_discharge_pressure: float,
+        rate_per_stream: NDArray[np.float64],
+        stream_to_maximize: int,
+    ) -> float:
+        """NB: Constraining to calculating maximum_rate for ingoing streams for now - the need is to figure out how much rate
+        can be added.
+
+        """
+        stream_to_maximize_connected_to_stage_no = self.streams[stream_to_maximize].connected_to_stage_no
+
+        # if it is not an ingoing stream --> currently no calculations done
+        # Fixme: what should be returned? 0.0, NaN or something else?
+        if not self.streams[stream_to_maximize].is_inlet_stream:
+            return 0.0
+
+        std_rates_std_m3_per_day_per_stream = rate_per_stream.copy()
+
+        def _calculate_train_result(std_rate_for_stream: float) -> CompressorTrainResultSingleTimeStep:
+            """Partial function of self.calculate_compressor_train_given_rate_ps_speed
+            where we only pass std_rate_per_stream and speed.
+            """
+            std_rates_std_m3_per_day_per_stream[stream_to_maximize] = std_rate_for_stream
+            return self.calculate_shaft_speed_given_rate_ps_pd(
+                std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
+
+        train_result = self.calculate_shaft_speed_given_rate_ps_pd(
+            std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+            suction_pressure=suction_pressure,
+            target_discharge_pressure=target_discharge_pressure,
+        )
+        if not train_result.is_valid:
+            zero_stream_result = _calculate_train_result(std_rate_for_stream=EPSILON)
+            if not zero_stream_result.is_valid:
+                return 0.0
+            else:
+                return maximize_x_given_boolean_condition_function(
+                    x_min=0.0,
+                    x_max=float(rate_per_stream[stream_to_maximize]),
+                    bool_func=lambda x: _calculate_train_result(std_rate_for_stream=x).is_valid,
+                )
+        else:
+            max_rate_is_larger_than = std_rates_std_m3_per_day_per_stream[stream_to_maximize]
+        while train_result.is_valid:
+            max_rate_is_larger_than = train_result.stage_results[
+                stream_to_maximize_connected_to_stage_no
+            ].standard_rate_asv_corrected_sm3_per_day
+            std_rates_std_m3_per_day_per_stream[stream_to_maximize] = max_rate_is_larger_than * 2
+            train_result = self.calculate_shaft_speed_given_rate_ps_pd(
+                std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
```

Comment:
Is this to find a rough max for the maximize function?

Reply:
We know what the rate should be larger than, so we double it until we reach something that is too large. Then we have an interval that we know is_valid will switch from true to false in - and we start looking for the point where it switches.

Reply:
What is the convergence criteria? EPSILON?

Reply:
The interval is halved and we continue in the upper or lower half depending on if the evaluation in the middle is true (upper half) or false (lower half). The evaluation stops when (highest value -lowest value)/highest value is less than 1e-5 or after 50 iterations.